sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -132,6 +130,15 @@ class Column(protected[sql] val expr: Expression) extends Logging {
     case _ => UnresolvedAttribute.quotedString(name)
   })
 
+  override def toString: String = usePrettyExpression(expr).sql
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Column => that.expr.equals(this.expr)
```

Comment:
how about `that.expr.semanticEquals(this.expr)`? One column equals to another if they always produce same result.